File: Python/3-SlidingWindow/Problems/2-LongestSubstringWithoutRepeatingCharacters/LongestSubstringWithoutRepeatingCharacters.py
'''
Sliding Window: Longest Substring Without Repeating Characters

Problem Statement:
Given a string s, find the length of the longest substring without repeating characters.
'''

# 1. Brute Force Approach (O(n^3))
def longest_unique_substring_brute_force(s):
    """
    Finds the length of the longest substring without repeating characters using a brute-force approach.
    It checks every possible substring for uniqueness.

    Args:
        s (str): The input string.

    Returns:
        int: The length of the longest unique substring.
    """
    max_length = 0  # Initialize the maximum length to 0.
    for i in range(len(s)):  # Iterate through all possible start indices of substrings.
        for j in range(i, len(s)):  # Iterate through all possible end indices for each start index.
            if len(set(s[i:j+1])) == (j - i + 1):  # Check if the substring s[i:j+1] has unique characters.
                # set(s[i:j+1]) creates a set of characters in the substring.
                # If the length of the set is equal to the length of the substring, all characters are unique.
                max_length = max(max_length, j - i + 1)  # Update max_length if the current substring is longer.
    return max_length  # Return the maximum length found.

# 2. Sliding Window with HashSet (O(n))
def longest_unique_substring_hashset(s):
    """
    Finds the length of the longest unique substring using a sliding window and a HashSet.
    The HashSet stores the characters in the current window.

    Args:
        s (str): The input string.

    Returns:
        int: The length of the longest unique substring.
    """
    seen = set()  # Initialize an empty HashSet to store characters in the current window.
    left = 0  # Initialize the left pointer of the sliding window.
    max_length = 0  # Initialize the maximum length to 0.
    for right in range(len(s)):  # Iterate through the string using the right pointer.
        while s[right] in seen:  # If the current character is already in the HashSet (i.e.,重复),
            seen.remove(s[left])  # Remove the leftmost character from the HashSet.
            left += 1  # Shrink the window from the left.
        seen.add(s[right])  # Add the current character to the HashSet.
        max_length = max(max_length, right - left + 1)  # Update max_length with the current window size.
    return max_length  # Return the maximum length.

# 3. Sliding Window with HashMap (O(n))
def longest_unique_substring_hashmap(s):
    """
    Finds the length of the longest unique substring using a sliding window and a HashMap.
    The HashMap stores characters and their indices.

    Args:
        s (str): The input string.

    Returns:
        int: The length of the longest unique substring.
    """
    char_index = {}  # Initialize an empty HashMap to store character indices.
    left = 0  # Initialize the left pointer of the sliding window.
    max_length = 0  # Initialize the maximum length to 0.
    for right in range(len(s)):  # Iterate through the string using the right pointer.
        if s[right] in char_index and char_index[s[right]] >= left:
            # If the current character is in the HashMap and its last seen index is within the current window:
            left = char_index[s[right]] + 1  # Move the left pointer to the position after the last occurrence of the character.
        char_index[s[right]] = right  # Update the index of the current character in the HashMap.
        max_length = max(max_length, right - left + 1)  # Update max_length.
    return max_length  # Return the maximum length.

# 4. Optimized Sliding Window with Array (O(n))
def longest_unique_substring_array(s):
    """
    Finds the length of the longest unique substring using a sliding window and an array.
    This approach is optimized for ASCII characters.

    Args:
        s (str): The input string.

    Returns:
        int: The length of the longest unique substring.
    """
    char_pos = [-1] * 128  # Initialize an array of size 128 (for ASCII characters) with -1.
    # char_pos[ord(c)] stores the last seen index of character c.
    left = 0  # Initialize the left pointer of the sliding window.
    max_length = 0  # Initialize the maximum length to 0.
    for right in range(len(s)):  # Iterate through the string using the right pointer.
        left = max(left, char_pos[ord(s[right])] + 1)  # Update left pointer.
        # ord(s[right]) gets the ASCII value of the character.
        # If the character has been seen before, char_pos[ord(s[right])] will be >= 0.
        # We move left to one position after the last occurrence.
        char_pos[ord(s[right])] = right  # Update the last seen position of the current character.
        max_length = max(max_length, right - left + 1)  # Update max_length.
    return max_length  # Return the maximum length.

# 5. Two Pointers with Frequency Map (O(n))
def longest_unique_substring_frequency_map(s):
    """
    Finds the length of the longest unique substring using two pointers and a frequency map (dictionary).

    Args:
        s (str): The input string.

    Returns:
        int: The length of the longest unique substring.
    """
    freq = {}  # Initialize an empty dictionary to store character frequencies.
    left = 0  # Initialize the left pointer.
    max_length = 0  # Initialize the maximum length.
    for right in range(len(s)):  # Iterate through the string using the right pointer.
        freq[s[right]] = freq.get(s[right], 0) + 1  # Increment the frequency of the current character.
        while freq[s[right]] > 1:  # If the frequency of the current character is greater than 1 (i.e., repeating),
            freq[s[left]] -= 1  # Decrement the frequency of the leftmost character.
            left += 1  # Move the left pointer to shrink the window.
        max_length = max(max_length, right - left + 1)  # Update max_length.
    return max_length  # Return the maximum length.

# Test Cases
s = "abcabcbb"
print("Brute Force:", longest_unique_substring_brute_force(s))
print("Sliding Window with HashSet:", longest_unique_substring_hashset(s))
print("Sliding Window with HashMap:", longest_unique_substring_hashmap(s))
print("Optimized Sliding Window with Array:", longest_unique_substring_array(s))
print("Two Pointers with Frequency Map:", longest_unique_substring_frequency_map(s))

File: C++/3-SlidingWindow/Problems/2-LongestSubstringWithoutRepeatingCharacters/LongestSubstringWithoutRepeatingCharacters.cpp
#include <iostream>
#include <string>
#include <vector>
#include <unordered_map>
#include <algorithm>
#include <limits>
#include <unordered_set> // Include the header for unordered_set

using namespace std;

// Function to find the length of the longest substring without repeating characters using the Brute Force approach.
// Time Complexity: O(n^3), Space Complexity: O(min(n, m)) where n is the length of the string and m is the size of the character set.
int longestUniqueSubstrBruteForce(const string& str) {
    int n = str.length();
    int longestLength = 0;

    // Iterate through all possible starting positions of substrings.
    for (int i = 0; i < n; i++) {
        // Iterate through all possible ending positions of substrings.
        for (int j = i; j < n; j++) {
            // Check if the substring from i to j has all unique characters.
            vector<bool> visited(256, false); // Assume ASCII character set

            bool unique = true;
            for (int k = i; k <= j; k++) {
                if (visited[str[k]]) {
                    unique = false;
                    break;
                }
                visited[str[k]] = true;
            }

            // If the substring has all unique characters, update the longest length.
            if (unique) {
                longestLength = max(longestLength, j - i + 1);
            }
        }
    }
    return longestLength;
}

// Function to find the length of the longest substring without repeating characters using the Sliding Window approach with a set.
// Time Complexity: O(n), Space Complexity: O(min(n, m))
int longestUniqueSubstrSlidingWindowSet(const string& str) {
    int n = str.length();
    int longestLength = 0;
    int left = 0, right = 0;
    unordered_set<char> charSet;

    while (right < n) {
        if (charSet.find(str[right]) == charSet.end()) {
            // If the character is not in the set, add it and move the right pointer.
            charSet.insert(str[right]);
            right++;
            longestLength = max(longestLength, (int)charSet.size());
        } else {
            // If the character is already in the set, remove the leftmost character and move the left pointer.
            charSet.erase(str[left]);
            left++;
        }
    }
    return longestLength;
}

// Function to find the length of the longest substring without repeating characters using the Sliding Window approach with a map.
// Time Complexity: O(n), Space Complexity: O(min(n, m))
int longestUniqueSubstrSlidingWindowMap(const string& str) {
    int n = str.length();
    int longestLength = 0;
    int left = 0, right = 0;
    unordered_map<char, int> charMap; // Stores character and its last seen index

    while (right < n) {
        if (charMap.find(str[right]) == charMap.end() || charMap[str[right]] < left) {
            // If the character is not in the map or its last seen index is before the current window,
            // add it to the map with its current index and move the right pointer.
            charMap[str[right]] = right;
            right++;
            longestLength = max(longestLength, right - left);
        } else {
            // If the character is in the map and its last seen index is within the current window,
            // move the left pointer to just after the last seen index of the repeating character.
            left = charMap[str[right]] + 1;
            charMap[str[right]] = right; // Update the last seen index
            right++;
        }
    }
    return longestLength;
}

// Function to find the length of the longest substring without repeating characters using the Optimized Sliding Window approach with a vector.
// Time Complexity: O(n), Space Complexity: O(m) where m is the size of the character set (e.g., 256 for ASCII).
int longestUniqueSubstrOptimized(const string& str) {
    int n = str.length();
    int longestLength = 0;
    int left = 0, right = 0;
    vector<int> charIndex(256, -1); // Stores the last seen index of each character, initialized to -1

    while (right < n) {
        if (charIndex[str[right]] == -1 || charIndex[str[right]] < left) {
            // If the character has not been seen or its last seen index is outside the current window
            charIndex[str[right]] = right; // Update last seen index
            right++; // Expand window
            longestLength = max(longestLength, right - left); // Calculate and store max length
        } else {
            // If the character is repeated within the window
            left = charIndex[str[right]] + 1; // Shrink window to exclude the repeated char
            charIndex[str[right]] = right;    // Update last seen index
            right++;
        }
    }
    return longestLength;
}

// Another optimized approach using only a vector
int longestUniqueSubstrVector(const string& str) {
    int n = str.length();
    if (n <= 1) return n;

    vector<int> lastIndex(256, -1); // Initialize a vector to store the last seen index of each character
    int maxLength = 0;
    int start = 0;

    for (int i = 0; i < n; ++i) {
        unsigned char currentChar = str[i]; // Use unsigned char to handle ASCII values correctly
        if (lastIndex[currentChar] != -1 && lastIndex[currentChar] >= start) {
            start = lastIndex[currentChar] + 1; // Move the start of the window
        }
        lastIndex[currentChar] = i; // Update the last seen index
        maxLength = max(maxLength, i - start + 1); // Calculate the maximum length
    }
    return maxLength;
}

int main() {
    string str = "abcabcbb";
    cout << "String: " << str << endl;

    cout << "\nBrute Force Approach: Length of longest unique substring: " << longestUniqueSubstrBruteForce(str) << endl;
    cout << "Sliding Window with Set: Length of longest unique substring: " << longestUniqueSubstrSlidingWindowSet(str) << endl;
    cout << "Sliding Window with Map: Length of longest unique substring: " << longestUniqueSubstrSlidingWindowMap(str) << endl;
    cout << "Optimized Sliding Window: Length of longest unique substring: " << longestUniqueSubstrOptimized(str) << endl;
    cout << "Optimized Sliding Window with Vector: Length of longest unique substring: " << longestUniqueSubstrVector(str) << endl;

    string str2 = "bbbbb";
    cout << "\nString: " << str2 << endl;
    cout << "\nBrute Force Approach: Length of longest unique substring: " << longestUniqueSubstrBruteForce(str2) << endl;
    cout << "Sliding Window with Set: Length of longest unique substring: " << longestUniqueSubstrSlidingWindowSet(str2) << endl;
    cout << "Sliding Window with Map: Length of longest unique substring: " << longestUniqueSubstrSlidingWindowMap(str2) << endl;
    cout << "Optimized Sliding Window: Length of longest unique substring: " << longestUniqueSubstrOptimized(str2) << endl;
    cout << "Optimized Sliding Window with Vector: Length of longest unique substring: " << longestUniqueSubstrVector(str2) << endl;

    string str3 = "pwwkew";
    cout << "\nString: " << str3 << endl;
    cout << "\nBrute Force Approach: Length of longest unique substring: " << longestUniqueSubstrBruteForce(str3) << endl;
    cout << "Sliding Window with Set: Length of longest unique substring: " << longestUniqueSubstrSlidingWindowSet(str3) << endl;
    cout << "Sliding Window with Map: Length of longest unique substring: " << longestUniqueSubstrSlidingWindowMap(str3) << endl;
    cout << "Optimized Sliding Window: Length of longest unique substring: " << longestUniqueSubstrOptimized(str3) << endl;
    cout << "Optimized Sliding Window with Vector: Length of longest unique substring: " << longestUniqueSubstrVector(str3) << endl;

    string str4 = "";
    cout << "\nString: " << str4 << endl;
    cout << "\nBrute Force Approach: Length of longest unique substring: " << longestUniqueSubstrBruteForce(str4) << endl;
    cout << "Sliding Window with Set: Length of longest unique substring: " << longestUniqueSubstrSlidingWindowSet(str4) << endl;
    cout << "Sliding Window with Map: Length of longest unique substring: " << longestUniqueSubstrSlidingWindowMap(str4) << endl;
    cout << "Optimized Sliding Window: Length of longest unique substring: " << longestUniqueSubstrOptimized(str4) << endl;
    cout << "Optimized Sliding Window with Vector: Length of longest unique substring: " << longestUniqueSubstrVector(str4) << endl;

    string str5 = "abcdefg";
    cout << "\nString: " << str5 << endl;
    cout << "\nBrute Force Approach: Length of longest unique substring: " << longestUniqueSubstrBruteForce(str5) << endl;
    cout << "Sliding Window with Set: Length of longest unique substring: " << longestUniqueSubstrSlidingWindowSet(str5) << endl;
    cout << "Sliding Window with Map: Length of longest unique substring: " << longestUniqueSubstrSlidingWindowMap(str5) << endl;
    cout << "Optimized Sliding Window: Length of longest unique substring: " << longestUniqueSubstrOptimized(str5) << endl;
    cout << "Optimized Sliding Window with Vector: Length of longest unique substring: " << longestUniqueSubstrVector(str5) << endl;

    return 0;
}

